### PR TITLE
[DebugInfo] Enable -fno-system-debug by default

### PR DESCRIPTION
**Pull request by LU-JOHN**
Enable -fno-system-debug by default.   Even if -fno-system-debug is enabled do, generate debug info to access deque's value_type.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -3844,10 +3844,12 @@ def fdebug_macro : Flag<["-"], "fdebug-macro">, Group<f_Group>,
 def fno_debug_macro : Flag<["-"], "fno-debug-macro">, Group<f_Group>,
   Visibility<[ClangOption, CLOption, DXCOption]>,
   HelpText<"Do not emit macro debug information">;
+def fsystem_debug : Flag<["-"], "fsystem-debug">, Group<f_Group>,
+  Visibility<[ClangOption, CC1Option, CLOption, DXCOption]>,
+  HelpText<"Emit debug information for declarations in system headers">;
 def fno_system_debug : Flag<["-"], "fno-system-debug">, Group<f_Group>,
```
Comment (asudarsa) [inline clang/include/clang/Driver/Options.td:3850]: Do we still need this option?

Thanks

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 2]
```diff
@@ -6056,8 +6056,19 @@ bool clang::CodeGen::noSystemDebugInfo(const Decl *D,
     if (isa<TypedefDecl>(D))
       return false;
 
-    // -fno-system-debug was used.  Do not generate debug info.
-    if (CGM.getCodeGenOpts().NoSystemDebug)
+#ifndef WINDOWS
+    // Generate debug for _M_insert_aux to allow debugging for
```
Comment (asudarsa) [inline clang/lib/CodeGen/CGDebugInfo.cpp:6060]: It is not clear what is being attempted here. Can you please add a comment? Thanks

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 3]
```diff
@@ -6056,8 +6056,19 @@ bool clang::CodeGen::noSystemDebugInfo(const Decl *D,
...
+    // deque's value_type.
+    const auto *Constructor = dyn_cast<CXXConstructorDecl>(D);
+    if (Constructor && Constructor->getNameAsString() == "deque")
+      return false;
+    const auto *Method = dyn_cast<CXXMethodDecl>(D);
+    if (Method && Method->getNameAsString() == "_M_insert_aux")
+      return false;
+#endif
+
+    // -fsystem-debug was not used.  Do not generate debug info.
```
Comment (asudarsa) [inline clang/lib/CodeGen/CGDebugInfo.cpp:6070]: Can be rewritten.

// Generate debug info only if -fsystem-debug is specified

File: clang/test/CodeGenCXX/debug-sin.cpp [hunk 4]
```diff
@@ -1,23 +1,28 @@
...
 //
-// Also ensure that no debug info for sin() is generated if -fno-system-debug is used.
+// Also ensure that no debug info for sin() is generated by default or if -fno-system-debug is used.
 
 // Debug info for math library functions is not generated on Windows
 // UNSUPPORTED: system-windows
 
+// RUN: %clang -fsystem-debug    -emit-llvm -S -g %s     -o %t.system_debug.ll
 // RUN: %clang                   -emit-llvm -S -g %s     -o %t.default.ll
 // RUN: %clang -fno-system-debug -emit-llvm -S -g %s     -o %t.no_system_debug.ll
```
Comment (asudarsa) [inline clang/test/CodeGenCXX/debug-sin.cpp:13]: I am not sure if we need testing with -fno-system-debug. Ideally, it will be great if we can remove fno-system-debug everything.

Thanks

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (LU-JOHN) [reply clang/include/clang/Driver/Options.td:3850]: Originally, we only had -fno-system-debug, but now -fno-system-debug is on by default, so -fsystem-debug is important to revert to old behavior.  -fsystem-debug is needed for the testcase linetable-eh.cpp

It could be argued that -fno-system-debug is no longer necessary, since it is on by default.  However, it is better to have on/off pairs for a boolean option.

File: clang/lib/CodeGen/CGDebugInfo.cpp (hunk 3 above)
Comment (LU-JOHN) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Changed.

File: clang/lib/CodeGen/CGDebugInfo.cpp (hunk 2 above)
Comment (LU-JOHN) [reply clang/lib/CodeGen/CGDebugInfo.cpp:6060]: Updated comment to explain why #ifndef WINDOWS is used.

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 5]
```diff
@@ -6056,8 +6056,21 @@ bool clang::CodeGen::noSystemDebugInfo(const Decl *D,
...
+    // Generate debug for _M_insert_aux to allow debugging for
+    // deque's value_type.
+    // _M_insert_aux is not used by Windows implementation, so
+    // there is no need for this check on Windows.
+    const auto *Constructor = dyn_cast<CXXConstructorDecl>(D);
+    if (Constructor && Constructor->getNameAsString() == "deque")
+      return false;
+    const auto *Method = dyn_cast<CXXMethodDecl>(D);
+    if (Method && Method->getNameAsString() == "_M_insert_aux")
+      return false;
```
Comment (premanandrao) [inline clang/lib/CodeGen/CGDebugInfo.cpp:6069]: What is unique about these two entries that they are special-cased?  Feels very hacky.

**Review by premanandrao — CHANGES_REQUESTED** [review]
I am really not fond of the function name special cases in CGDebugInfo.cpp

**Review by elizabethandrews — COMMENTED** [review]
This change feels very odd to me. Why are we turning off system debug info by default and why are we excluding the cases in the PR when we are turning them off?

File: clang/lib/CodeGen/CGDebugInfo.cpp (hunk 5 above)
Comment (LU-JOHN) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Remove special case to preserve _M_insert_aux.  Try to find clean solution later.  Small loss of debuggability can be dealt with by using -fsystem-debug.

**Comment by LU-JOHN** [comment]
> This change feels very odd to me. Why are we turning off system debug info by default and why are we excluding the cases in the PR when we are turning them off?

We are excluding system debug info by default because it gives a large reduction in debug build size, sometimes 4x.  Build time is also reduced, sometimes by ~50%.   There is almost no loss in debuggability.  1 or 2 cases out of >10K cases.

**Comment by stevemerr** [comment]
If I understand your comment here:  
https://jira.devtools.intel.com/browse/CMPLRLLVM-24314?focusedId=22284384&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-22284384 
this option was approved as being off by default.  Why is it being enabled by default now?

This will certainly have a negative impact on debugging and is not recommended as the default compiler behavior by the DGI team.

**Comment by LU-JOHN** [comment]
> If I understand your comment here: https://jira.devtools.intel.com/browse/CMPLRLLVM-24314?focusedId=22284384&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-22284384 this option was approved as being off by default. Why is it being enabled by default now?
> 
> This will certainly have a negative impact on debugging and is not recommended as the default compiler behavior by the DGI team.

Initially "no system debug" was guarded by an option, but our testing with the gdb-14.1 testsuite and the gdbCPP testsuite found very few debug issues.  1 out of >10K testcases.  This issue is understood, and we have a workaround that can be added to the release notes.  We feel the improvements to the issues customers are encountering with large debug file size and debug build time are worth the corner case debugging failure.

**Comment by stevemerr** [comment]
> > > If I understand your comment here:  this option was approved as being off by default. Why is it being enabled by default now?
> > > This will certainly have a negative impact on debugging and is not recommended as the default compiler behavior by the DGI team.
> > 
> > 
> > Initially "no system debug" was guarded by an option, but our testing with the gdb-14.1 testsuite and the gdbCPP testsuite found very few debug issues. 1 out of >10K testcases. This issue is understood, and we have a workaround that can be added to the release notes. We feel the improvements to the issues customers are encountering with large debug file size and debug build time are worth the corner case debugging failure.
> 
> System headers change over time. It is by design that these test suites doesn't try to verify debugging of code coming from sources outside that which can be controlled by the test. The impact to these test suites would not be an appropriate gauge of end user impact.

To help clarify, I'll give some specific examples of how this change could impact an end user.  Using the nestedtypes_01.cpp sample included in this changeset as an example, the user has declared a local variable C10Seq within main().  The user would typically expect to have complete debug info for this local variable however some of the differences a user would see using gdb after this change:

- Examining the type of C10seq (ptype C10seq) loses info
- Stepping into calls like C10seq.push_back, C10seq.begin or C10seq.end is no longer possible
- Stopping at a breakpoint set in one of the above routines will no longer display parameters or source code

**Comment by LU-JOHN** [comment]
> To help clarify, I'll give some specific examples of how this change could impact an end user. Using the nestedtypes_01.cpp sample included in this changeset as an example, the user has declared a local variable C10Seq within main(). The user would typically expect to have complete debug info for this local variable however some of the differences a user would see using gdb after this change:
> 
> * Examining the type of C10seq (ptype C10seq) loses info
> * Stepping into calls like C10seq.push_back, C10seq.begin or C10seq.end is no longer possible
> * Stopping at a breakpoint set in one of the above routines will no longer display parameters or source code

Thanks for your examples.  I verified that with -fno-system-debug stepping into push_back is unsurprisingly not possible and setting a breakpoint at stl_deque.h:1509 is not possible.   When I tried ptype and print on Cl0Seq I saw the same result with/without -fno-system-debug.

```
type = std::deque<Cl0>
$1 = std::deque with 10 elements = {{mX = 0}, {mX = 1}, {mX = 2}, {mX = 3}, {mX = 4}, {mX = 5}, {mX = 6}, {mX = 7}, {mX = 8}, {mX = 9}}
```

**Comment by stevemerr** [comment]
> > To help clarify, I'll give some specific examples of how this change could impact an end user. Using the nestedtypes_01.cpp sample included in this changeset as an example, the user has declared a local variable C10Seq within main(). The user would typically expect to have complete debug info for this local variable however some of the differences a user would see using gdb after this change:
> > 
> > * Examining the type of C10seq (ptype C10seq) loses info
> > * Stepping into calls like C10seq.push_back, C10seq.begin or C10seq.end is no longer possible
> > * Stopping at a breakpoint set in one of the above routines will no longer display parameters or source code
> 
> Thanks for your examples. I verified that with -fno-system-debug stepping into push_back is unsurprisingly not possible and setting a breakpoint at stl_deque.h:1509 is not possible. When I tried ptype and print on Cl0Seq I saw the same result with/without -fno-system-debug.
> 
> ```
> type = std::deque<Cl0>
> $1 = std::deque with 10 elements = {{mX = 0}, {mX = 1}, {mX = 2}, {mX = 3}, {mX = 4}, {mX = 5}, {mX = 6}, {mX = 7}, {mX = 8}, {mX = 9}}
> ```

Not sure why we see different results for ptype, perhaps you have a pretty printer enabled?  You can likely see the difference if you try "ptype std::deque<Cl0, std::allocator<Cl0> >" instead.  Below is the expanded type info I see in gdb:

```
(gdb) ptype Cl0Seq
type = class std::deque<Cl0, std::allocator<Cl0> > [with _Tp = Cl0, 
         _Alloc = std::allocator<Cl0>] 
        : protected std::_Deque_base<_Tp, _Alloc> {

  public:
    typedef std::_Deque_base<_Tp, _Alloc>::iterator iterator;
}
```

**Comment by LU-JOHN** [comment]
It was decided that -fno-system-debug should not be enabled by default.
